urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/json/rpc/JSONRPCHandler.java

```diff
@@ -0,0 +1,802 @@
+	    
+	    return this.createParamClassArrayFromJSONObject( request.getParams().asJSONObject() );
+
+	} else {
+	    throw createJSONRPCException( "params must be an array or an object." );	    
+	}
+    }
+
+
+    /**
+     * This method builds up the parameter class array from the passed
+     * JSON array param.
+     *
+     * This type mapping will be used:
+     *   - JSON number:   Integer or Double
+     *   - JSON boolean:  Boolean
+     *   - JSON string:   String
+     *   - JSON null:     Object
+     *   - JSON array:    JSONArray
+     *   - JSON object:   JSONObject
+     *
+     * Since version 1.0.1 all JSON param types (including JSONArray and 
+     * JSONObject) are supported in addition to String, Number, Boolean and Null.
+     *
+     * @param params A JSON array containing the method call params.
+     * @return A class-array matching the passed params (returned in the same order).
+     * @throws JSONRPCException 
+     **/
+    private Class<?>[] createParamClassArrayFromJSONArray( JSONArray params ) 
+	throws JSONRPCException, 
+	       JSONException {
+
+	// List<JSONValue> paramList = request.getParams().getArray();
+	Class<?>[] paramClasses = new Class<?>[ params.getArray().size() ];
+
+	for( int i = 0; i < params.getArray().size(); i++ ) {
+	    JSONValue param = params.getArray().get(i);
+
+	    if( !this.unwrapJSONValues() )
+		paramClasses[i] = JSONValue.class;
+	    else if( param.isNull() )
+		paramClasses[i] = Object.class;
+	    else if( param.isBoolean() )
+		paramClasses[i] = Boolean.class;
+	    else if( param.isNumber() ) 
+		paramClasses[i] = param.getNumber().getClass(); // Double or Integer
+	    else if( param.isString() )
+		paramClasses[i] = String.class;
+	    else if( param.isArray() )
+		paramClasses[i] = JSONArray.class;
+	    else if( param.isObject() )
+		paramClasses[i] = JSONObject.class;
+	    else 
+		throw new JSONRPCException( "datatype at param " + i +" is not supported in the param list." );
+	}
+	
+	return paramClasses;
+
+    }
+	
+    /**
+     * Using JSON objects as params DOES NOT WORK!
+     *
+     * See http://stackoverflow.com/questions/2237803/can-i-obtain-method-parameter-name-using-java-reflection
+     **/
+    private Class<?>[] createParamClassArrayFromJSONObject( JSONObject params ) 
+	throws JSONException,
+	       JSONRPCException {
+
+
+	// Check param types
+	// PROBLEM:
+	//  - getting parameter names is possible if debug information is included during compilation. See this answer for more details
+	//  - otherwise getting parameter names is not possible
+	// See http://stackoverflow.com/questions/2237803/can-i-obtain-method-parameter-name-using-java-reflection
+
+	// It MIGHT be possible to convert the object into an array, IF the
+	// member names are numbers that address the array indices.
+	try {
+	    
+	    return this.createParamClassArrayFromJSONArray( params.asJSONArray() );
+
+	} catch( JSONException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)